### PR TITLE
HUB-720 Working verify-local-startup

### PR DESCRIPTION
**Pull request by j-maynard**
Comes with working variant-c support.  Generates its own certificates and runs first time every time.  Some highlight:

* Updated to a more recent verison of ruby
* Dockerfile has been updated to use more recent versions of Go and Ruby
* Moved urls.env to sit in the same place as ports.env.
* Removed fpm from the Gemfile as it doesn't appear to be used and is prevent us from using a more recent version of the JSON gem.  The older version was throwing deprecation warnings.
* URLs have been removed from ports.env, config/env.sh now sources from urls.env
* local-fed-config has been updated and now works
* docker-compose has received numerous updates and now works
* repos.yml now points all hub builds to build from local.Dockerfile

*Dependancies*

Depends on changes in:

- [ ] verify-hub
- [x] verify-frontend (now merged)
- [x] verify-matching-service-adapter (now merged).

*Known Bugs*
- [ ] Stub IDP is currently missing its logos and some images
- [ ] Only 3 stub-idps currently show up in the IDP pick page

### CODE REVIEW

File: configuration/stub-frontend-fed-config/ab_test.yml [hunk 1]
```diff
@@ -9,31 +9,45 @@
...
           percent: 0
-  - loa1_perf_management:
+  - short_hub_2019_q3:
       alternatives:
-        - name: 'control'
-          percent: 100
-        - name: 'variant'
+        - name: 'control_a'
           percent: 0
+        - name: 'variant_c_2_idp_short_hub'
```
Comment (CharlesIC) [inline configuration/stub-frontend-fed-config/ab_test.yml:48]: Short hub experiment is going away so do we need this for local hub? It's probably better to just make it run with control.
My concern is that this file will need to be updated every time we have an AB test, otherwise local running may break. It would be better to make it independent of AB tests or use the AB test config file from verify-frontend instead of replicating it. I think ida-hub-acceptance tests works that way.

**Review by CharlesIC — COMMENTED** [review]
Could this be made independent of changes to AB tests or use config from verify-frontend? More like ida-hub-acceptance-tests works

File: configuration/stub-frontend-fed-config/ab_test.yml (hunk 1 above)
Comment (j-maynard) [reply configuration/stub-frontend-fed-config/ab_test.yml:48]: I can investigate this...  I think to meet the intentions of this repo which is to allow anyone to spin up the hub it has its own version of verify-hub-frontend-federation-config which is also where the files relating to the AB test normally live.  Let me do a quick investigation over lunchtime.

Comment (j-maynard) [reply configuration/stub-frontend-fed-config/ab_test.yml:48]: Updated this to point to the one in verify-frontend repository.  Also updated the file to match what is currently in production to stop it breaking while we complete the AB test tear down work.  (See PR: https://github.com/alphagov/verify-frontend/pull/904)

File: configuration/stub-frontend-fed-config/idp-rules-variant-c/stub-idp-no-docs.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+simpleIds: [ "stub-idp-no-docs" ]
```
Comment (jkunle) [inline configuration/stub-frontend-fed-config/idp-rules-variant-c/stub-idp-no-docs.yml:1]: You want to look at https://github.com/alphagov/verify-frontend-federation-config/compare/HUB-737-tear-down-shorthub-abtest-and-keep-variant to adjust the rules to match

File: configuration/stub-frontend-fed-config/special-cases/abc-variants.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+---
```
Comment (jkunle) [inline configuration/stub-frontend-fed-config/special-cases/abc-variants.yml:1]: The special-cases folder will no longer be needed, I and John have removed this see https://github.com/alphagov/verify-frontend-federation-config/commit/8e1dfee16e077efe63936ec2d3bdd165a6a35db0

File: configuration/stub-frontend-fed-config/idp-rules-variant-c/stub-idp-no-docs.yml (hunk 2 above)
Comment (jkunle) [reply configuration/stub-frontend-fed-config/idp-rules-variant-c/stub-idp-no-docs.yml:1]: Generally this is good stuff, and would probably just need a few tweaks once the remove short-hub is done.

File: configuration/stub-frontend-fed-config/special-cases/abc-variants.yml (hunk 3 above)
Comment (j-maynard) [reply configuration/stub-frontend-fed-config/special-cases/abc-variants.yml:1]: Could we make a list of these changes and then can we raise a ticket?  I accept you're making breaking changes to the frontend and how it's configured but until those land I can't make the changes here without breaking verify-local-startup.  The other ideal is we land this and then you and John can make the adjustments and submit a PR ;-)

File: configuration/stub-frontend-fed-config/idp-rules-variant-c/stub-idp-no-docs.yml (hunk 2 above)
Comment (j-maynard) [reply configuration/stub-frontend-fed-config/idp-rules-variant-c/stub-idp-no-docs.yml:1]: Yeah (see other comment) I agree... There is either a follow up ticket or we land this so you and John can make the changes as part of your ticket.  I am flexible either way.  I am keen to land this as it is now so this can stop being broken.  It might also help you and John with this now working.

File: urls.env [hunk 4]
```diff
@@ -1,7 +1,7 @@
 METADATA_URL=http://metadata/dev.xml
-POLICY_URL=http://policy
-CONFIG_URL=http://config
-SAML_PROXY_URL=http://saml-proxy
+POLICY_URL=http://policy:50110
+CONFIG_URL=http://config:50240
+SAML_PROXY_URL=http://saml-proxy:50220
```
Comment (instantiator) [inline urls.env:4]: We have ports for some of these, but not others - but also we have a config/ports file. Let's talk about how it all fits together.

**Review by instantiator — APPROVED** [review]

File: generate/generate-metadata.sh [hunk 5]
```diff
@@ -49,16 +49,12 @@ for src in dev compliance-tool; do
...
   
   # sign
-  if test -z `which xmlsectool`; then
-    echo "$(tput setaf 3)Installing xmlsectool$(tput sgr0)"
-    brew install xmlsectool
-  fi
-  
   echo "$(tput setaf 3)Signing metadata$(tput sgr0)"
-  xmlsectool \
+  $XMLSECTOOL \
```
Comment (Wynndow) [inline generate/generate-metadata.sh:53]:
```suggestion
  xmlsectool \
```
I think you can leave this as is. The value gets conditionally set on line 9, so should pick up the ENV variable if set, or default to just `xmlsectool`

File: generate/generate-metadata.sh [hunk 6]
```diff
@@ -49,16 +49,12 @@ for src in dev compliance-tool; do
   fi
   
   # sign
-  if test -z `which xmlsectool`; then
```
Comment (Wynndow) [inline generate/generate-metadata.sh:52]: Why was this removed?

File: urls.env (hunk 4 above)
Comment (Wynndow) [reply urls.env:4]: I'm wondering a similar thing - what was the resolution?

**Review by Wynndow — COMMENTED** [review]
I don't know much about the ab_test stuff but have a few comments about the ports etc.

File: config/ports.env [hunk 7]
```diff
@@ -1,6 +1,4 @@
-METADATA_PORT=55500
-METADATA_URL=http://localhost:55500/dev.xml
-
+METADATA_PORT=80
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply config/ports.env:1]: So METADATA in verify-local-startup is actually an Nginx instance and I've not had much luck getting it to run on anything other than port 80... So internally within the docker network it runs on 80, externally you can access it directly on port 55500 to see the XML and grab the certs for debugging.

File: urls.env (hunk 4 above)
Comment (j-maynard) [reply urls.env:4]: So these are fully qualified URLs which are used in the config.  The ports in the ports.env file tell the services generally what ports to run on and are sometimes used to configure the opening of ports in the docker-compose.yml

File: configuration/stub-frontend-fed-config/idp-rules/stub-idp-three.yml [hunk 8]
```diff
@@ -1,11 +1,49 @@
-simpleIds: [ "stub-idp-three" ]
+simpleIds: ["stub-idp-demo-three" ]
 
 segments:
   protected:
-    recommended: []
+    recommended: ["2doc_gbdl_pp_mobile",
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply configuration/stub-frontend-fed-config/idp-rules/stub-idp-three.yml:5]: Because they didn't work when i tested them.

File: docker-compose.yml [hunk 9]
```diff
@@ -1,12 +1,22 @@
...
 
 services:
+  redis:
+    image: redis:5.0.6
+    volumes:
+      - redis-data:/data
+      - redis-config:/usr/local/etc/redis/
+    command: /bin/sh -c 'redis-server'
+    ports:
+      - 6379:6379
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply docker-compose.yml:11]: I can add the REDIS port to ports.env that said its not made available outside of the docker network.

File: generate/generate-metadata.sh (hunk 6 above)
Comment (j-maynard) [reply generate/generate-metadata.sh:52]: because this runs with in a docker container where we are guaranteed to have  xmlsectool... Also the the following code targeted homebrew which doesn't run so well on linux or in docker.

File: config/ports.env (hunk 7 above)
[the remark this answers is not in the document]
Comment (Wynndow) [reply config/ports.env:1]: Yep, agreed. As is though this variable doesn't get used - in the docker compose the internal port is hard coded to port 80. So I think it might be better to keep this var as 55500 and use it instead of the new var created.

File: generate/generate-metadata.sh (hunk 6 above)
Comment (Wynndow) [reply generate/generate-metadata.sh:52]: Might be nice to leave it in though in case someone wants to run it without docker though?

Comment (j-maynard) [reply generate/generate-metadata.sh:52]: Then I think it should just dump out that it needs xmlsectool and maybe information on how to get it rather than trying to run brew.

Comment (Wynndow) [reply generate/generate-metadata.sh:52]: Yep, fair do's.

File: config/ports.env [hunk 10]
```diff
@@ -15,6 +13,12 @@ VSP_PORT=50400
 FRONTEND_PORT=50300
 STUB_IDP_PORT=50140
 
+TEST_RP_EXTERNAL_PORT=50130
+STUB_IDP_EXTERNAL_PORT=50140
+FRONTEND_EXTERNAL_PORT=50300
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply config/ports.env]: I have removed the first two... The external port is there so you can access the NGINX instance directly.

File: config/ports.env [hunk 11]
```diff
@@ -15,6 +13,12 @@ VSP_PORT=50400
...
+METADATA_EXTERNAL_PORT=55500
+MSA_EXTERNAL_PORT=50400
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply config/ports.env]: I have removed most of these... Other than the METADATA_EXTERNAL_PORT.

File: docker-compose.yml [hunk 12]
```diff
@@ -74,13 +135,15 @@ services:
...
+    ports:
+      - "${METADATA_EXTERNAL_PORT}:80"
     volumes:
       - ./data/metadata:/usr/share/nginx/html:ro
 
   stub-idp:
     image: ${STUB_IDP_IMAGE}
     ports:
-      - "${STUB_IDP_EXTERNAL_PORT}:80"
+      - "${STUB_IDP_EXTERNAL_PORT}:${STUB_IDP_EXTERNAL_PORT}"
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply docker-compose.yml]: I have fixed this.

File: docker-compose.yml [hunk 13]
```diff
@@ -89,22 +152,31 @@ services:
...
-      PORT: 80
+      PORT: ${STUB_IDP_EXTERNAL_PORT}
       LOG_PATH: /app/logs
       GRAPHITE_REPORTING_FREQUENCY: 24h
       STUB_IDP_BASIC_AUTH: "false"
       STUB_IDPS_FILE_PATH: /idps/stub-idps.yml
       ENTITY_ID: http://%s.local/SSO/POST
       STUB_IDP_ASSET_LOCATION: /assets
       STUB_IDP_YAML_FILE_LOCATION: /idps/stub-idps.yml
+      DB_URI: jdbc:postgresql://stub-idp-db:5432/postgres?user=postgres&password=docker
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply docker-compose.yml]: I have updated this to get it from the URLS.env file.

File: docker-compose.yml [hunk 14]
```diff
@@ -74,13 +135,15 @@ services:
 
   metadata:
     image: nginx
+    ports:
+      - "${METADATA_EXTERNAL_PORT}:80"
```
[the remark this answers is not in the document]
Comment (j-maynard) [reply docker-compose.yml:139]: This is the external port for NGINX to be access by the user for debugging.  So I have left this here.

File: generate/generate-metadata.sh (hunk 6 above)
Comment (j-maynard) [reply generate/generate-metadata.sh:52]: I have put this back in at line 15 - 20 to exit if XMLSecTool is missing.

File: generate/generate-metadata.sh (hunk 5 above)
Comment (j-maynard) [reply generate/generate-metadata.sh:53]: Set as a value... Which in bash has to be accessed as a variable ($VAR)

**Review by instantiator — APPROVED** [review]

**Review by idrop — APPROVED** [review]
